src/circuit_proof/r1cs.rs

```diff
@@ -53,112 +58,138 @@ impl LinearCombination {
+    aL_assignment: Vec<Assignment>,
+    aR_assignment: Vec<Assignment>,
+    aO_assignment: Vec<Assignment>,
+    v_assignment: Vec<Assignment>,
 }
 
 impl ConstraintSystem {
     pub fn new() -> Self {
         ConstraintSystem {
-            a: vec![],
-            b: vec![],
-            c: vec![],
-            var_assignment: vec![],
+            constraints: vec![],
+            aL_assignment: vec![],
+            aR_assignment: vec![],
+            aO_assignment: vec![],
+            v_assignment: vec![],
         }
     }
-    // Allocate a variable and do value assignment at the same time
-    pub fn alloc_variable(&mut self, val: Scalar) -> Variable {
-        self.var_assignment.push(val);
-        Variable(self.var_assignment.len() - 1)
-    }
 
-    // Push one set of linear constraints (a, b, c) to the constraint system.
-    // Pushing a, b, c together prevents mismatched constraints.
-    pub fn constrain(
+    // Allocate variables for left, right, and output wires of multiplication,
+    // and assign them the Result values that are passed in.
+    // Prover will pass in Ok(Scalar)s, and Verifier will pass in R1CSErrors.
+    pub fn assign_multiplier(
         &mut self,
-        lc_a: LinearCombination,
-        lc_b: LinearCombination,
-        lc_c: LinearCombination,
-    ) -> Result<(), &'static str> {
-        // TODO: check that the linear combinations are valid
-        // (e.g. that variables are valid, belong to this constraint system).
-        self.a.push(lc_a);
-        self.b.push(lc_b);
-        self.c.push(lc_c);
-        Ok(())
+        left: Assignment,
+        right: Assignment,
+        out: Assignment,
+    ) -> (Variable, Variable, Variable) {
+        self.aL_assignment.push(left);
+        let left_var = Variable::MultiplierLeft(self.aL_assignment.len() - 1);
+
+        self.aR_assignment.push(right);
+        let right_var = Variable::MultiplierRight(self.aR_assignment.len() - 1);
+
+        self.aO_assignment.push(out);
+        let out_var = Variable::MultiplierOutput(self.aO_assignment.len() - 1);
+
+        (left_var, right_var, out_var)
     }
 
-    fn eval_lc(&self, lc: &LinearCombination) -> Scalar {
-        let sum_vars: Scalar = lc
-            .variables
-            .iter()
-            .map(|(var, scalar)| scalar * self.var_assignment[var.0])
-            .sum();
-        sum_vars + lc.constant
+    // Allocate a committed variable, and assign it the Result value passed in.
+    // Prover will pass in Ok(Scalar), and Verifier will pass in R1CSError.
+    pub fn assign_committed_variable(&mut self, value: Assignment) -> Variable {
+        self.v_assignment.push(value);
+        Variable::Committed(self.v_assignment.len() - 1)
     }
 
-    fn create_verifier_input(
+    pub fn multipliers_count(&self) -> usize {
+        let n = self.aL_assignment.len();
+        if n == 0 || n.is_power_of_two() {
+            return n;
+        }
+        return n.next_power_of_two();
+    }
```

Comment:
Maybe this should be
```
fn innerproduct_size(&self) -> usize {
    self.aL_assignment.len().next_power_of_two()
}
```
since what this function returns isn't actually the number of multipliers, but the size of the inner product vectors?

Except for the `n == 0` case, it's not necessary to check whether `n` is already a power of two, and I think we can probably exclude `n == 0` in this new setting because `n` will be zero only when we have an empty constraint system, which shouldn't (imo) be allowed.